daras_ai_v2/base.py

```diff
@@ -1594,23 +1647,57 @@ def render_extra_waiting_output(self):
     def estimate_run_duration(self) -> int | None:
         pass
 
-    def submit_and_redirect(self):
+    def submit_and_redirect(self) -> typing.NoReturn | None:
         sr = self.on_submit()
         if not sr:
             return
         raise gui.RedirectException(self.app_url(run_id=sr.run_id, uid=sr.uid))
 
+    def publish_and_redirect(self) -> typing.NoReturn | None:
+        assert self.request.user and not self.request.user.is_anonymous
+
+        if (
```

Comment:
This transaction stuff is more complex than it needs to be. You can do something like:

```py
matches = SavedRun.objects.filter(
    id=CURRENT_RUN.ID,
    run_status="", error_msg="", run_time=datetime.timedelta()  # filter for pending state
).update(run_status=STARTING_STATE, uid=LOGGED_IN_USER.UID)
if matches > 0:  # no matches, run has probably started elsewhere
    >> send task to queue, create published run & redirect
else:
    >> rerun
```

Reply:
thanks